circuits/circuits/distribute/README.md

```diff
@@ -0,0 +1,48 @@
+# Circuits
+
+As well as being used to enforce the state transition function, ZKPs are used to handle distribution after shutdown.
+The distribution circuits take a set of claims over coins and output the rightful final owners given those claims.
+
+## Algorithm
+
+Distribution involves 3 parts:
+    - Filter out any claim that is invalid (i.e., did not occur in the state history)
+    - Resolve any conflicts between claims
+    - Prove that there are no conflicting claims in the final output
+
+For conceptual simplicity and good prover time, we will implement each part as a separate Nova circuit.
+
+### Filtering
+
+Claims are statments like `X owns coins [a,b] in block Y` along with a purported Merkle proof.
+The set of claims are represented as an accumulator, in particular a keccak hash chain.
+The state history is represented as a Merkle root, which is known onchain, and passed as a public input to the circuit. It must be passed unchanged through each iteration of the circuit.
+There is also an index public input, which is required to start at `0` and end at `claim_count`, which is also known onchain.
+There is a `filtered_count` input that starts at `0`.
+
+The circuit processes one claim per iteration/fold. The claim accumulator starts at 0, one claim is added per iteration, and it must end with the right onchain value.
+If the claim exists in the state history, the `filtered_count` variable is incremented and the claim is added to a filtered poseidon merkle tree at position `filtered_count`. Otherwise, `filtered_count` stays the same and nothing is changed.
+
+If all the checks pass, then this circuit outputs a `filtered_tree` of size `filtered_count`, where every claim in the tree is valid.
+
+### Challenge Resolution
+
+The challenge resolution circuit is also in Nova and runs iteratively. The initial input is the `filtered_tree` which is a poseidon Merkle tree, and each iteration takes two contradictory claims in the tree, resolves the contradiction, and updates the tree.
+
+The first claim is called the winner and the second is the loser. The winner's block number must be greater than the loser's. The winner's leaf remains unedited. If the winner's coins are a superset of the loser's (i.e., $[0,5]$ beats $[2,3]$), the loser's leaf is deleted, and the `resolved_count` is decremented. If the loser's coins are split (i.e., $[2,3]$ beats $[0,5]$), the loser's leaf is deleted, but two new leaves are added and the `resolved_count` is incremented. If the loser's coins are truncated (i.e., $[1,5]$ vs $[1,3]$), the loser's leaf is deleted, a new leaf is added and the `resolved_count` remains the same.
```

Comment:
> loser's coins are truncated (i.e., $[1,5]$ vs $[1,3]$)

do you mean $[1,5]$ vs $[1,6]$? the provided example looks a superset to me.

Reply:
Yup good catch!